File: 4 kyu/Human readable duration format/solution.cpp
#include <string>
#include <vector>

std::string format_duration(int seconds) {
  if (seconds == 0) return "now";
  
  std::vector<std::string> times;
  auto add = [&] (auto text, auto time) {
    if (time == 0) return;
    times.push_back(std::to_string(time) + text + (time > 1 ? "s" : ""));
  };
  add(" year",    seconds / 31536000);    
  add(" day",    (seconds / 86400) % 365);    
  add(" hour",   (seconds / 3600) % 24);    
  add(" minute", (seconds / 60) % 60);    
  add(" second",  seconds % 60);    
  
  std::string result = times[0];
  for (auto i = 1; i < times.size() - 1; i ++) result.append(", " + times[i]);
  if (times.size() > 1) result.append(" and " + times.back());
  
  return result;
}

___________________________________________________
#include <iostream>
#include <string>
#include <cmath>
using namespace std;
std::string format_duration(int seconds) {
int gs = seconds;
  int years = gs/(3.154e7);
  int days = (gs-years*31536000)/86400;
int ngs = gs - ((24 * 3600 * days) + (31536000 * years));
  int hours = (ngs / 3600);
int nngs = ngs - (3600 * hours);
  int minutes = nngs / 60;
int nnngs = nngs - (60 * minutes);
  int secs = nnngs;
string strdays = to_string(days);
string strhours = to_string(hours);
string strminutes = to_string(minutes);
string strsecs = to_string(secs);
string stryears = to_string(years);
string sy, sd, sh, sm, ss;
  if (years == 0 && days == 0 && hours == 0 && minutes == 0 && secs == 1)
    {return "1 second";}
    else if (years == 0 && days == 0 && hours == 0 && minutes == 1 && secs == 0)
      {return "1 minute";}
    else if (years == 0 && days == 0 && hours == 1 && minutes == 0 && secs == 0)
      {return "1 hour";}
    else if (years == 0 && days == 1 && hours == 0 && minutes == 0 && secs == 0)
      {return "1 day";}
    else if (years == 1 && days == 1 && hours == 0 && minutes == 0 && secs == 0)
      {return "1 year";}
    else if (years == 0 && days == 0 && hours == 0 && minutes == 0 && secs == 0)
      {return "now";}
  
  if (years == 1) {sy = "1 year";}
    else if (years>1) {sy = stryears + " years";}
  if (days == 1) {sd = "1 day";}
    else if (days>1) {sd = strdays + " days";}
  if (hours == 1) {sh = "1 hour";}
    else if (hours>1) {sh = strhours + " hours";}
  if (minutes == 1) {sm = "1 minute";}
    else if (minutes>1) {sm = strminutes + " minutes";}
  if (secs == 1) {ss = "1 second";}
    else if (secs>1) {ss = strsecs + " seconds";}

  if (years == 0 && days == 0 && hours == 0 && minutes != 0 && secs != 0)
    {sm = sm + " and ";}
    else if (years == 0 && days == 0 && hours != 0 && minutes != 0 && secs != 0)
      {sh = sh + ", "; sm = sm + " and ";}
    else if (years == 0 && days != 0 && hours != 0 && minutes != 0 && secs != 0)
      {sd = sd + ", "; sh = sh + ", "; sm = sm + " and ";}
    else if (years != 0 && days != 0 && hours != 0 && minutes != 0 && secs != 0)
      {sy = sy + ", "; sd = sd + ", "; sh = sh + ", "; sm = sm + " and ";}
    else if (years == 0 && days == 0 && hours == 0 && minutes == 0 && secs == 0)
      {return "1 year";}
    else if (years == 0 && days != 0 && hours == 0 && minutes != 0 && secs != 0)
      {sd = sd + ", "; sm = sm + " and ";}
    else if (years == 0 && days != 0 && hours == 0 && minutes != 0 && secs != 0)
      {sd = sd + ", "; sm = sm + " and ";}
    else if (years == 0 && days != 0 && hours != 0 && minutes == 0 && secs != 0)
      {sd = sd + ", "; sh = sh + " and ";}
    else if (years == 0 && days != 0 && hours != 0 && minutes != 0 && secs == 0)
      {sd = sd + ", "; sh = sh + " and ";}
    else if (years == 0 && days != 0 && hours == 0 && minutes == 0 && secs != 0)
      {sd = sd + " and ";}
    else if (years == 0 && days != 0 && hours != 0 && minutes == 0 && secs == 0)
      {sd = sd + " and ";}
    else if (years == 0 && days == 0 && hours != 0 && minutes != 0 && secs == 0)
      {sh = sh + " and ";}
    else if (years == 0 && days != 0 && hours == 0 && minutes != 0 && secs == 0)
      {sd = sd + " and ";}
    else if (years != 0 && days != 0 && hours != 0 && minutes != 0 && secs == 0)
      {sy = sy + ", "; sd = sd + ", "; sh = sh + " and ";}
  if (years == 0 && days == 0 && hours == 0 && minutes == 0 && secs == 0)
      {return "now";}
  return sy + sd + sh + sm + ss;
}

___________________________________________________
#include <string>
#include <vector>

std::string format_duration(int s) {
    if(s == 0){ return "now"; }
    std::vector<std::pair<std::string, int>> data =
    {{"year",(s/60/60/24/365)},{"day",(s/60/60/24)%365},{"hour",(s/60/60)%24},{"minute",(s/60)%60},{"second",s%60}};
    std::vector<std::string> tvec;
    for(auto &c:data){
        if(c.second != 0){
            tvec.push_back(std::to_string(c.second) + " " + c.first + (c.second > 1? "s":""));
        }
    }
    std::string result;
    size_t size = tvec.size();
    for(int i = 0; i != size; i++){
        result += tvec[i] + (i==size-1?(""):(i == size-2?(" and "):(", ")));
    }
    return result;
}

File: 4 kyu/Human readable duration format/solution.rs
fn format_duration(seconds: u64) -> String {
    let result = [
        ("year", 31536000, 100000),
        ("day", 86400, 365),
        ("hour", 3600, 24),
        ("minute", 60, 60),
        ("second", 1, 60),
    ].iter()
    .map(|(unit, duration, modulo)| (seconds / duration % modulo, unit))
    .filter_map(|(duration, unit)| match duration {
        _ if duration == 1 => Some(format!("{} {}", duration, unit)),
        _ if duration != 0 => Some(format!("{} {}s", duration, unit)),
        _ => None,
    }).collect::<Vec<String>>();

    match result.len() {
        0 => String::from("now"),
        1 => result.join(""),
        _ => result
            .split_last()
            .map(|(r, l)| l.join(", ") + " and " + r)
            .unwrap(),
    }
}

__________________________________________________
const SECOND: u64 = 1;
const MINUTE: u64 = 60 * SECOND;
const HOUR: u64 = 60 * MINUTE;
const DAY: u64 = 24 * HOUR;
const YEAR: u64 = 365 * DAY;

enum Unit {
    SECONDS,
    MINUTES,
    HOURS,
    DAYS,
    YEARS,
}

struct Time {
    kind: Unit,
    amt: u64,
}

impl Time {
    fn unit_str(&self) -> &str {
        match self.kind {
            Unit::SECONDS => "second",
            Unit::MINUTES => "minute",
            Unit::HOURS => "hour",
            Unit::DAYS => "day",
            Unit::YEARS => "year",
        }
    }
    
    fn as_string(&self) -> String {
        let has_s = |x| if x > 1 { "s" } else { "" };
        
        format!("{} {}{}", self.amt, self.unit_str(), has_s(self.amt))
    }
}

fn format_duration(seconds: u64) -> String {
    let years = seconds / YEAR;
    let days = (seconds % YEAR) / DAY;
    let hours = (seconds % DAY) / HOUR;
    let minutes = (seconds % HOUR) / MINUTE;
    let secs = seconds % MINUTE;
    
    let t = |u, l| Time{ kind: u, amt: l };
    
    let times = 
        [t(Unit::YEARS, years), t(Unit::DAYS, days), t(Unit::HOURS, hours), 
            t(Unit::MINUTES, minutes), t(Unit::SECONDS, secs)]
        .iter()
        .filter(|x| x.amt > 0)
        .map(|t| t.as_string())
        .collect::<Vec<String>>();
    
    match times.len() {
        0 => String::from("now"),
        1 => times.first().unwrap().to_string(),
        _ => format!("{} and {}", times[..times.len() - 1].join(", "), times.last().unwrap()),
    }
}

__________________________________________________
fn format_duration(seconds: u64) -> String {
    let y = seconds / 31536000;
    let d = seconds / 86400 % 365;
    let h = seconds / 3600 % 24;
    let m = seconds / 60 % 60;
    let s = seconds % 60;
    let time: [(u64, &'static str); 5] = [(y, "year"), (d, "day"), (h, "hour"), (m, "minute"), (s, "second")];
    let str_v: Vec<_> = time.into_iter().filter(|t| t.0 != 0)
        .map(|&(i, u)| format!("{} {}{}", i, u, if i == 1 { "" } else { "s" })).collect();
    match str_v.len() {
        0 => "now".into(),
        1 => str_v[0].clone(),
        c => str_v.iter().cloned().take(c - 1).collect::<Vec<_>>().join(", ") + &format!(" and {}", str_v[c - 1]),
    }
}
